### PR TITLE
fix: prevent using backslash in file names (v3)

### PR DESCRIPTION
**Pull request by fzipi**
Prevents using backslash in file names.

Fixes 3MU-240701-2 for v3.

Thanks to @luelueking for reporting this.

### CODE REVIEW

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml [hunk 1]
```diff
@@ -103,3 +103,97 @@ tests:
...
+              Content-Type: multipart/form-data; boundary=---------------------------265001916915724
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+            method: POST
+            port: 80
+            uri: /
+            data: |
+              -----------------------------265001916915724
+              Content-Disposition: form-data; name="test"; filename="1.j\s\p"
+              Content-Type: application/octet-stream
+
```
Comment (fzipi) [inline tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:124]:
```suggestion

```

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml [hunk 2]
```diff
@@ -103,3 +103,96 @@ tests:
...
+              Host: "localhost"
+              Content-Type: multipart/form-data; boundary=---------------------------265001916915724
+              Accept: "text/xml,application/xml,application/xhtml+xml,text/html;q=0.9,text/plain;q=0.8,image/png,*/*;q=0.5"
+            method: POST
+            port: 80
+            uri: /
+            data: |
+              -----------------------------265001916915724
+              Content-Disposition: form-data; name="test"; filename="1.j\s\p"
+              Content-Type: application/octet-stream
```
Comment (fzipi) [inline tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:123]:
```suggestion
              Content-Type: application/octet-stream

```

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml [hunk 3]
```diff
@@ -103,3 +103,96 @@ tests:
...
+              Host: "localhost"
+              Content-Type: multipart/form-data; boundary="--abcdefg"
+              Accept: "*/*"
+            method: POST
+            port: 80
+            uri: /
+            data: |
+              --abcdefg
+              Content-Disposition: form-data; name="t\e\st.t\x\t"
+              Content-Type: application/octet-stream
```
Comment (fzipi) [inline tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:170]:
```suggestion
              Content-Type: application/octet-stream

```

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml [hunk 4]
```diff
@@ -103,3 +103,96 @@ tests:
...
```
Comment (fzipi) [inline tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:193]:
```suggestion
              Content-Type: application/octet-stream

```

**Review by fzipi — COMMENTED** [review]
Fix spacing in test.

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml [hunk 5]
```diff
@@ -103,3 +103,51 @@ tests:
...
+              Content-Disposition: form-data; name="test"; filename="1.j\s\p"
+              Content-Type: application/octet-stream
+
+              helloworld
+              -----------------------------265001916915724--
+            version: "HTTP/1.1"
+          output:
+            log_contains: id "920120"
+  - test_title: 920120-5
+    desc: "Attempted multipart/form-data bypass (920121). Match \\ in FILES_NAMES."
```
Comment (theseion) [inline tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:131]:
```suggestion
    desc: 'Attempted multipart/form-data bypass (920121). Match \ in FILES.'
```

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml [hunk 6]
```diff
@@ -103,3 +103,51 @@ tests:
               -----------------------------265001916915724--
           output:
             log_contains: id "920120"
+  - test_title: 920120-4
+    desc: "Attempted multipart/form-data bypass (920120). Match \\ in FILES_NAMES."
```
Comment (theseion) [inline tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:107]:
```suggestion
    desc: 'Attempted multipart/form-data bypass (920120). Match \ in FILES.'
```

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml
Comment (theseion) [inline tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:1]: Add tests for `FILES_NAMES`.

**Review by theseion — CHANGES_REQUESTED** [review]

File: tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml
Comment (fzipi) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:1]: I discussed this with @airween a bit a couple days ago. Let me copy his thoughts here.

Using these rules as example:
```apacheconf
SecRule FILES "@rx .*" \
    "id:1000,\
    phase:2,\
    capture,\
    msg:'%{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
    allow"

SecRule FILES_NAMES "@rx .*" \
    "id:1001,\
    phase:2,\
    capture,\
    msg:'%{MATCHED_VAR_NAME}: %{MATCHED_VAR}',\
    allow"
```
and this payload:
```
--a
Content-Disposition: form-data; name="file";filename="1.jsp"

<%out.print(123)%>
--a--
```
When tested, these are the logs:
```
[Wed Aug 21 21:14:22.490736 2024] [security2:error] [pid 509880:tid 509882] [client ::1:52968] ModSecurity: Warning. Pattern match ".*" at FILES:file. [file "/home/airween/src/coreruleset/rules/temp.conf"] [line "468"] [id "1000"] [msg "FILES:file: 1.jsp"] [hostname "localhost"] [uri "/post"] [unique_id "ZsY8jnwUPHCjdbq0IiOjiwAAAAA"]
[Wed Aug 21 21:14:22.490873 2024] [security2:error] [pid 509880:tid 509882] [client ::1:52968] ModSecurity: Warning. Pattern match ".*" at FILES_NAMES:file. [file "/home/airween/src/coreruleset/rules/temp.conf"] [line "475"] [id "1001"] [msg "FILES_NAMES:file: file"] [hostname "localhost"] [uri "/post"] [unique_id "ZsY8jnwUPHCjdbq0IiOjiwAAAAA"]
```
So as you can see, `FILES` is a collection, where the contents of `name` are in the key and the contents of `filename` in the value. `FILES_NAMES` is also a collection, but it seems that [only carries the names of the variables that _contain a file_](https://github.com/owasp-modsecurity/ModSecurity/wiki/Reference-Manual-(v2.x)#user-content-FILES_NAMES). So for this case, it is unusable as variable.

Does this sound reasonable?

TL; DR: we should remove `FILES_NAMES` from this rule probably, and create a new issue to see which other rules include this variable and evaluate if we have benefits on the inclusion.

Comment (theseion) [reply tests/regression/tests/REQUEST-920-PROTOCOL-ENFORCEMENT/920120.yaml:1]: Yes, `FILES_NAMES` contains the names of the form fields. However, I can see an application using the form field name as a canonical file name. Consider a form where you upload a CV. The field name could be `cv.pdf`. This is certainly contrived but it's possible. If an application would use this method, then an attacker could simply change the filename by changing the name of the form field.

I think it would simply be safer to use both targets.

**Review by theseion — APPROVED** [review]
